grails-app/services/au/org/emii/aatams/detection/DetectionExtractService.groovy

```diff
@@ -59,39 +57,48 @@ class DetectionExtractService extends AbstractStreamingExporterService {
     protected void writeCsvData(final filterParams, OutputStream out)
     {
         filterParams.sql = new Sql(dataSource)
+        filterParams.sql.fetchSize = getFetchSize()
+        filterParams.sql.autoCommit = false
         filterParams.projectPermissionCache = [:]
 
         super.writeCsvData(filterParams, out)
     }
 
-    protected def readData(filterParams)
+    protected def eachRow(filterParams, closure)
     {
-        return extractPage(filterParams)
+        def startTime = System.currentTimeMillis()
+        def query =  new QueryBuilder().constructQuery(filterParams)
+        def releaseIsProtectedCache = [:]
+        int count = 0
+        
+        filterParams.sql.eachRow(query.getSQL(), query.getBindValues()) { row ->
+            def sanitisedRows = applyEmbargo([row.toRowResult()], filterParams, releaseIsProtectedCache)
+            sanitisedRows.each { sanitisedRow ->
+                closure.call(sanitisedRow)
+            }
+            count++
+        }
```

Comment:
I haven't looked into it, @jonescc, but just one thing that strikes me here is that we are using a method that takes a list (`applyEmbargo()`) but always passing it a one-element list. Also we call `sanitisedRows.each{ ... }` but that will just be an array with one or no elements. I guess my question is just whether there is any function from the SQL class which could let us process all the rows as a group rather than one-by-one. Not a huge deal but there just seems to be one extra level of nesting which we might be able to avoid.